5: Rental Screen

chriswith8 opened the pull request.

FrancescoGhisi committed:
fix: Inicial config

FrancescoGhisi committed:
fix: addMouseListener toolsTable (seleciona item na tabela)

fix: addActionListener lendToolButton (Verifica se a ferramenta já está emprestada, se não, abre a tela de empréstimo.

FrancescoGhisi committed:
fix: addMouseListener toolsTable (seleciona item na tabela)

fix: addActionListener lendToolButton (Verifica se a ferramenta já está emprestada, se não, abre a tela de empréstimo.

in progress: combo box para mostrar a lista de amigos, o botão de empréstimo e label para mostrar o nome da ferramenta.

chriswith8 commented on src/main/java/tool/rental/presentation/RegisterUserFrame.java:
```diff
@@ -15,7 +15,7 @@ public class RegisterUserFrame extends PresentationFrame {
     private JPasswordField passwordField;
     private JButton confirmButton;
-    private JButton cancelButton;
+    private JButton cancelButton; // Chris, favor configurar o botão!!
```
Eu não! Abram issue!

FrancescoGhisi replied:
Feito!
![Issue - Cancel Button](https://github.com/chrislcontrol/tool-rental/assets/146891303/a6f4795b-e823-42d5-aca1-777cce183c16)

chriswith8 commented on src/main/java/tool/rental/domain/repositories/RentalRepository.java:
```diff
@@ -50,4 +50,24 @@ public void updateDevolutionTimestamp(Rental rental, long devolutionTimestamp) t
         }
     }
+
+    public boolean verifyRental(String idSelecionado ) throws ToastError{
```
```suggestion
    public boolean isRentalOpen(String idSelecionado ) throws ToastError{
```

FrancescoGhisi replied:
Done.

chriswith8 commented on src/main/java/tool/rental/presentation/AppMainFrame.java:
```diff
@@ -30,6 +33,7 @@ public class AppMainFrame extends PresentationFrame {
 
     public AppMainFrame() throws ToastError {
         this.setMainPanel();
+        this.lendToolButton.setEnabled(false);
```
Para manter o projeto mais simples, vamos deixar o botão como enabled e só não fazer nada caso a ferramenta não esteja selecionada e o usuário clicar no botão. Similar com o que foi feito em "devolver ferramenta"

FrancescoGhisi replied:
Done.

chriswith8 commented on src/main/java/tool/rental/presentation/AppMainFrame.java:
```diff
@@ -102,6 +106,22 @@ private void setMainPanel() {
 
+//EM TESTE
+
+    private String idSelecionado;
```
Esse código, até a linha 124, não deveria estar aqui. Ao invés disso, na hora de instanciar o outro frame, deve ser passada a ferramenta no construtor do LendToolRental. Assim, dentro da classe java de LendToolRental, vc terá acesso à ferramenta que foi selecionada.

FrancescoGhisi replied:
Feito.

chriswith8 commented on src/main/java/tool/rental/presentation/AppMainFrame.java:
```diff
@@ -114,6 +134,36 @@ public void actionPerformed(ActionEvent e) {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                try {
+                    if(rentalRepository.verifyRental(getIdSelecionado())){
```
Essa verificação deve ficar dentro do UseCase

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    public ResultSet listFriends() throws ToastError{
```
Método deve estar dentro do FriendRepository

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,92 @@
+                        SELECT name FROM FRIEND ORDER BY name
+                        """;
+            PreparedStatement stm = dataBase.connection.prepareStatement(query);
+            return stm.executeQuery();
```
```suggestion
            return db.executeQuery(stm);
```

chriswith8 replied:
Ao invés de retornar o ResultSet da query, deve retornar List<Friend>

Commit:
[5] removed idea

FrancescoGhisi committed:
fix: removed method setEnabled lendToolButton;
fix: renamed method verifyRental to isRentalOpen.

FrancescoGhisi committed:
Fix: Removido o atributo 'idSelecionado';
Fix: Método actionPerformed do lendToolButton;
Fix: Removido o método setEnabled do rentButton;

FrancescoGhisi committed:
Fix: Moved listFriend method from LendToolFrame.java to FriendRepository.java

Fix: listFriend method in progress...

FrancescoGhisi committed:
Fix: changed variable name from 'idSelecionado' to 'toolId'

FrancescoGhisi committed:
Feat: Added friend's name into friendList ComboBox in LendToolFrame.java

FrancescoGhisi committed:
Feat: Added verification if tool is rented.

FrancescoGhisi committed:
Fix: Resolved merge conflicts

FrancescoGhisi committed:
Fix: Removed else in openRegisterRentalModel method

FrancescoGhisi committed:
Merge branch 'master' of https://github.com/chrislcontrol/tool-rental into 5

FrancescoGhisi committed:
Merge branch 'master' of https://github.com/chrislcontrol/tool-rental into 5

FrancescoGhisi committed:
Fix: change ComboBox

FrancescoGhisi committed:
Fix: Merge conflicts

FrancescoGhisi committed:
Fix: Table in LendToolFrame

Shoto29 committed:
Merge remote-tracking branch 'origin/5' into 5

# Conflicts:
#	src/main/java/tool/rental/presentation/FriendsScreenFrame.java

Shoto29 committed:
Update: Change in the visual profile of lentToolFrame screen

FrancescoGhisi committed:
JLtoolName

FrancescoGhisi committed:
Merge fix

FrancescoGhisi committed:
Fix: Merge Conflicts

FrancescoGhisi committed:
WIP

Commit:
[5] fix: wrong timestamp format

FrancescoGhisi committed:
Feat: rentButton on LendToolFrame is configured. When it's clicked: Register rent to tool; Exception if user will rent the selected tool for another friend; It returns to AppMainFrame and it updates toolsTable.

FrancescoGhisi committed:
Fix: Merge conflicts

FrancescoGhisi committed:
Feat: added filter name to friendsTable on LendToolFrame

FrancescoGhisi committed:
Feat: Added input dialog to receive the user's confirmation to rent a selected tool;
Feat: Added an advice to show to user if selected friend already has a tool rented, when user clicks on rentButton.

chriswith8 commented on src/main/java/tool/rental/domain/controllers/AppMainController.java:
```diff
@@ -94,6 +96,26 @@ public void returnTool(String toolId) throws ToastError {
     }
 
+    public void openRegisterRentalModal(String toolId, String toolName) throws ToastError {
+        if (isRentalOpenUseCase.isToolRented(toolId)) {
```
Use cases so devem ter um método publico: o execute

chriswith8 commented on src/main/java/tool/rental/domain/controllers/LendToolFrameController.java:
```diff
@@ -0,0 +1,40 @@
+    private final RentalToolUseCase rentalToolUseCase = new RentalToolUseCase();
+    private final IsRentalOpenUseCase isRentalOpenUseCase = new IsRentalOpenUseCase();
+
+    public LendToolFrameController(LendToolFrame Frame){
```
```suggestion
    public LendToolFrameController(LendToolFrame frame){
```

chriswith8 commented on src/main/java/tool/rental/domain/controllers/LendToolFrameController.java:
```diff
@@ -0,0 +1,40 @@
+    private final IsRentalOpenUseCase isRentalOpenUseCase = new IsRentalOpenUseCase();
+
+    public LendToolFrameController(LendToolFrame Frame){
+        super(Frame);
```
frame

chriswith8 commented on src/main/java/tool/rental/domain/controllers/LendToolFrameController.java:
```diff
@@ -0,0 +1,40 @@
+       return isRentalOpenUseCase.isToolRented(toolId);
+    }
+
+    public void swapFrame(PresentationFrame nextFrame){
```
Esse método não precisa ser sobrescrito. User o da classe pai

chriswith8 commented on src/main/java/tool/rental/domain/controllers/LendToolFrameController.java:
```diff
@@ -0,0 +1,40 @@
+        this.rentalToolUseCase.execute(friend, tool);
+    }
+
+    public boolean isToolRented(String toolId) throws ToastError {
```
Esse método não é usado

FrancescoGhisi replied:
É usado para a segunda verificação se a ferramenta já foi emprestada, quando está na tela de empréstimo. Caso o usuário burle a volta para a MainFrame, esse método pode ser passado direto do usercase para o frame ou passa pelo controller?

chriswith8 commented on src/main/java/tool/rental/domain/repositories/FriendRepository.java:
```diff
@@ -77,6 +77,62 @@ public ArrayList<Friend> listAll() throws ToastError {
+                        LEFT JOIN USER u on f.user_id = u.id
+                        LEFT JOIN RENTAL r on r.friend_id = f.id and r.devolution_timestamp is null
+                        LEFT JOIN TOOL t on t.id = f.id
+                        WHERE f.id = ?
```
```suggestion
                        WHERE f.id = ? f.user_id = ?
```

FrancescoGhisi replied:
Alterado para WHERE f.id = ? AND f.user_id = ?

chriswith8 commented on src/main/java/tool/rental/domain/repositories/FriendRepository.java:
```diff
@@ -77,6 +77,62 @@ public ArrayList<Friend> listAll() throws ToastError {
+                            r.devolution_timestamp as r__devolution_timestamp
+                            
+                        FROM FRIEND f
+                        LEFT JOIN USER u on f.user_id = u.id
```
Não precisa dar join em User, sempre será o User de Settings.getUser()

chriswith8 commented on src/main/java/tool/rental/domain/repositories/FriendRepository.java:
```diff
@@ -77,6 +77,62 @@ public ArrayList<Friend> listAll() throws ToastError {
+                return null;
+            }
+
+            User user = new User(
```
Usar o User de Settings.getUser()

chriswith8 commented on src/main/java/tool/rental/domain/repositories/FriendRepository.java:
```diff
@@ -77,6 +77,62 @@ public ArrayList<Friend> listAll() throws ToastError {
+                            
+                        FROM FRIEND f
+                        LEFT JOIN USER u on f.user_id = u.id
+                        LEFT JOIN RENTAL r on r.friend_id = f.id and r.devolution_timestamp is null
```
Pra que está dando join no Rental? Ele não é usado aqui

chriswith8 commented on src/main/java/tool/rental/domain/repositories/RentalRepository.java:
```diff
@@ -51,4 +54,24 @@ public void updateDevolutionTimestamp(Rental rental, long devolutionTimestamp) t
         }
     }
 
+    public void updateRentalTimestamp(long rentalTimestamp, Friend friend, Tool tool) throws ToastError {
```
```suggestion
    public void create(long rentalTimestamp, Friend friend, Tool tool) throws ToastError {
```

chriswith8 commented on src/main/java/tool/rental/domain/use_cases/IsRentalOpenUseCase.java:
```diff
@@ -0,0 +1,34 @@
+
+public class IsRentalOpenUseCase {
+
+    public boolean isToolRented(String toolId ) throws ToastError {
```
Isso deve ficar dentro do ToolRepository

chriswith8 commented on src/main/java/tool/rental/domain/use_cases/RentalToolUseCase.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private boolean isAnyToolRentedByFriend(Friend friend) throws ToastError {
+        try (DataBase db = new DataBase()) {
```
Isso deve ficar dentro de ToolRepository

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,184 @@
+        cancelButton.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                LendToolFrame.this.dispose();
```
```suggestion
                close();
```

FrancescoGhisi replied:
Alterado. Criado o método closeframe() no controller, conforme feito também no controller da RegisterTool

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,184 @@
+                            null,
+                            "Empréstimo realizado com sucesso!"
+                            );
+                    lendToolFrameController.swapFrame(new AppMainFrame());
```
Swaps devem ser chamados dentro do controller

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,184 @@
+                    }
+                    String friendId = friendsTable.getValueAt(friendsTable.getSelectedRow(), 0).toString();
+                    lendToolFrameController.rentTool(friendId, toolId);
+                    JOptionPane.showMessageDialog(
```
Colocar esse dialog dentro do controller

chriswith8 commented on src/main/java/tool/rental/presentation/LendToolFrame.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    private void setupTable() throws ToastError {
+        DefaultTableModel model = (DefaultTableModel) this.friendsTable.getModel();
```
Usar TableConfigurator pra criar tabela. Na master, existem exemplos de como usar, se espelha lá. Se não tiver no seu código, dá pull na master

chriswith8 commented on src/main/java/tool/rental/presentation/RegisterUserFrame.java:
```diff
@@ -15,7 +15,7 @@ public class RegisterUserFrame extends PresentationFrame {
     private JPasswordField passwordField;
     private JButton confirmButton;
-    private JButton cancelButton;
+    private JButton cancelButton; // Chris, favor configurar o botão!!
```
Remover comentário

FrancescoGhisi committed:
Fix: Reformatted code

FrancescoGhisi committed:
Fix: Merge conflicts

FrancescoGhisi committed:
Fix: bugs

FrancescoGhisi committed:
Fix: Reformatted code
